### PR TITLE
Blobkzg Circuit for EIP 4844

### PR DESCRIPTION
**Pull request by mabbamOG**
### Description

This PR adds the circuit required to integrate Taiko with EIP 4844. Basically it takes data stored in the circuit and uses it for function evaluation. This is then compared with the blob opening checked in the smart contract with EIP 4844. Make sure the data fed into this circuit is checked to be the same as all the other claims of the proof.

### Issue Link

[[_link issue here_]](https://github.com/taikoxyz/zkevm-circuits/issues/37)

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

Contain the new circuit as well as dependencies which support the latest traits along with the latest halo2curves (hence the breaking change...but the circuit has not been integrated into the rest of the zkEVM so everything compiles for now)

### Rationale

Required for the L2 to make use of EIP 4844 Blobs, greatly improving costs for publishing new proofs

### How Has This Been Tested?

Unit tests (more to come)

### CODE REVIEW

File: zkevm-circuits/src/blobkzg_circuit/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+            integer_chip: &IntegerChip<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>,
+            ctx: &mut RegionCtx<N>,
+            x: W
+        ) -> Result<AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>, Error> {
+            let rc_rns: std::rc::Rc<Rns<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>> = std::rc::Rc::new(Rns::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::construct());
+            let i: Integer<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB> = Integer::from_fe(x, rc_rns);
+            let v: Value<Integer<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>> = Value::known(i);
+            let u: UnassignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB> = UnassignedInteger::from(v);
+            // TODO: is Range::Operand a correct range here?
+            let a: AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB> = integer_chip.assign_integer(ctx, u, Range::Operand)?;
```
Comment (xiaodino) [inline zkevm-circuits/src/blobkzg_circuit/src/main.rs:157]: What is `a` for?

File: zkevm-circuits/src/blobkzg_circuit/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,407 @@
+use halo2_proofs::plonk::{ConstraintSystem,Circuit, Error};
+use halo2_proofs::circuit::{Layouter,SimpleFloorPlanner,Value};
+use halo2_proofs::halo2curves::ff::PrimeField as FieldExt;
+use halo2wrong_integer::{IntegerConfig, rns::Rns, IntegerChip, IntegerInstructions, rns::Integer, UnassignedInteger, Range, AssignedInteger};
+use halo2wrong_maingate::{RangeChip, MainGate, RegionCtx};
```
Comment (xiaodino) [inline zkevm-circuits/src/blobkzg_circuit/src/main.rs:5]: What is the motivation to use `halo2wrong_maingate` here?

File: zkevm-circuits/src/blobkzg_circuit/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+                    }
+                }
+                nbits_bool
+            };
+
+            for (x, b) in nbits.iter_mut().zip(nbits_bool.iter()) {
+                *x = if *b {W::ONE} else {W::ZERO};
+            }
+            nbits
+        };
```
Comment (AlekseiVambol) [inline zkevm-circuits/src/blobkzg_circuit/src/main.rs:71]: Unfortunately, I do not no Halo2Wrong API and do not sure that I have correctly understood the related parts. However, if they work as I suppose and bit factorization of n (log2 of the order of w) is predefined, than I think that it should work correctly. The selected part of the code can be written in a shorter way (defining nbits[j*8+i] directly without the boolean array), but it is not a problem.

File: zkevm-circuits/src/blobkzg_circuit/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+        let x2is: [W; 256] = {
+            let mut x2is = [W::default(); 256];
+            let mut state = *x;
+            for x in &mut x2is {
+                *x = state;
+                state *= state;
+            }
+            x2is
+        };
+        let nbits: [W; 256] = {
```
Comment (AlekseiVambol) [inline zkevm-circuits/src/blobkzg_circuit/src/main.rs:54]: Unfortunately, I do not no Halo2Wrong API and do not sure that I have correctly understood the related parts. However, if they work as I suppose and bit factorization of n (log2 of the order of w) is predefined, than I think that it should work correctly. This part of the code (the lines 54-71) can be written in a shorter way (defining nbits[j*8+i] directly without the boolean array), but it is not a problem.

File: zkevm-circuits/src/blobkzg_circuit/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        const NUMBER_OF_LIMBS: usize = 4;
+
+        let integer_chip = {
+            let rc_rns = std::rc::Rc::new(Rns::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::construct());
+            IntegerChip::<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::new(config.integer_config, rc_rns)
+        };
+        use halo2wrong_maingate::RangeInstructions;
+        integer_chip.range_chip().load_table(&mut layouter)?;
+
+        layouter.assign_region(|| "BlobKZGCircuit Region", | region| {
```
Comment (Brechtpd) [inline zkevm-circuits/src/blobkzg_circuit/src/main.rs:182]: Some more comments here on what does what would also be nice.

**Review by Brechtpd — COMMENTED** [review]
Please use `cargo fmt` to format the code. Yeah I know it messes up the code how you want it to look, but it does make it easier to see the code in a nice standard way.

It seems like some tests did not pass as expected:
```
running 6 tests
test tests::wrong_witness_data ... FAILED
test tests::wrong_x - should panic ... ok
test tests::try_default has been running for over 60 seconds
test tests::try_random_datapoints has been running for over 60 seconds
test tests::wrong_n has been running for over 60 seconds
test tests::wrong_w has been running for over 60 seconds
test tests::try_default ... ok
test tests::wrong_n - should panic ... FAILED
test tests::wrong_w - should panic ... FAILED
test tests::try_random_datapoints ... ok

failures:

---- tests::wrong_witness_data stdout ----
thread 'tests::wrong_witness_data' panicked at 'not implemented', src/main.rs:519:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::wrong_n stdout ----
note: test did not panic as expected
---- tests::wrong_w stdout ----
note: test did not panic as expected

failures:
    tests::wrong_n
    tests::wrong_w
    tests::wrong_witness_data

test result: FAILED. 3 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out; finished in 144.90s
```

`wrong_witness_data` makes sense how it is now (but please change it so all tests pass even if still to do), but the others seems like unexpected.

**Comment by AlekseiVambol** [comment]
An important remark about the correct way of choosing the pseudorandom evaluation point x for the polynomial encoding the data. The PR is aimed at implementing protocol on the screenshot. 
image

Note that the first 2 steps insist on choosing the value of x as the hash of **both** the polynomial outside the circuit and its evaluations inside the circuit. This means that (in the case of one-phase synthesis) the circuit must contain the constraints representing the hashing of the data polynomial evaluations. This inner hash value should be passed as a public input to the circuit to make the result of the in-circuit hashing accessible to the verifier. When a smart contract verifies the proof for the circuit, it will be given this inner hash and use it along with the hash of the polynomial outside the circuit to compute the evaluation point x, which is also a public input.

Why do we need this inner hashing? To force the prover to choose the value of x strictly after the data polynomial evaluations in the circuit are fixed. If prover is allowed to change these evaluations after x is chosen, he can interpolate the polynomial, which is not equal to the one outside the circuit, but coincide with it at x. He can do it with polynomial time and space complexity and is free to choose the values at any 4095 of 4096 evaluation points!

How could we simplify this approach? I think we can avoid writing the hash constraints, if we have a tricky proving system that allows two-phase synthesis, has the challenge API (returning the hash of the witnesses committed at the first phase) and allows us to generate some public input data between the phases of the synthesis. In this case the challenge value can be used instead of the aforesaid inner hash value - it can be used (along with the public input part representing the hash of the polynomial outside the circuit) to compute x between the synthesis phases . The value of x will be added to the public input before the second phase of synthesis. Unfortunately, it seems that Halo2 does not support this dynamic public input generation.

**Comment by Brechtpd** [comment]
@AlekseiVambol Could we reuse the KZG commitment for the column storing the data that could already be available for the verifier (?) instead of explicitly committing to the data in-circuit?

**Comment by AlekseiVambol** [comment]
Yes.

**Pull request closed without merging**